common/lib/xmodule/xmodule/js/src/problem/edit.coffee

```diff
@@ -189,31 +192,129 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
     else
       return template
 
-# We may wish to add insertHeader. Here is Tom's code.
-# function makeHeader() {
-#  var selection = simpleEditor.getSelection();
-#  var revisedSelection = selection + '\n';
-#  for(var i = 0; i < selection.length; i++) {
-#revisedSelection += '=';
-#  }
-#  simpleEditor.replaceSelection(revisedSelection);
-#}
-#
+
   @markdownToXml: (markdown)->
     toXml = `function (markdown) {
       var xml = markdown,
           i, splits, scriptFlag;
 
+      // fix DOS \r\n line endings to look like \n
+      xml = xml.replace(/\r\n/g, '\n');
+
       // replace headers
       xml = xml.replace(/(^.*?$)(?=\n\=\=+$)/gm, '<h1>$1</h1>');
       xml = xml.replace(/\n^\=\=+$/gm, '');
 
-      // group multiple choice answers
+      // Pull out demand hints,  || a hint ||
+      var demandhints = '';
+      xml = xml.replace(/(^\s*\|\|.*?\|\|\s*$\n?)+/gm, function(match) {  // $\n
```

Comment:
Looks like this code is trying to transform Markdown to XML using regular expressions, rather than actually parsing the Markdown. That will work for awhile -- and judging by the fact that this function has been doing it this way already, it's already been working for awhile. However, at a certain point, this system is going to break down. Regular expressions are not designed for transforming documents from one system to another, and complicated Markdown is going to break this regex system.

The real solution here is to move to a system where the Markdown is parsed to an abstract syntax tree, which can then be re-rendered as XML. And I found a Javascript library that does just that: [markdown-js](https://github.com/evilstreak/markdown-js). However, since we're already using regular expressions, that is almost certainly outside of the scope of this pull request. I wanted to leave this comment here to make sure people are aware that markdown-js exists, and as a reminder to come back here at some point and re-write this Markdown-to-XML system using a proper parsing system.